nmea_parser.c

```diff
@@ -1,191 +1,315 @@
 #include "nmea_parser.h"
 #include "def.h"
 #include "utils.h"
+#include <sys/types.h>
+#include <sys/socket.h>
 
 extern int g_debug;
 extern int g_foreground;
 extern FILE *fp_console;
+static unsigned char status=0x26;
 
-void parse_NMEA_sensor(char* message, t_sensor_context* sensors)
+// Return a 0 if checksum is bad
+// Return the location of the '*' character if checksum is good.
+int verify_checksum (char *ptr) {
+	int i,j,k;
+
+	for (i=j=0;((ptr[i]!=0) && (ptr[i]!='*'));++i) j^=ptr[i];
+	k=j&0xf;
+	if (ptr[i]!='*') return 0;
+	if (k<10) {
+		if (ptr[i+2] != '0'+k) return 0;
+	} else { 
+		if (ptr[i+2] != 'A'+k-10) return 0;
+	}
+	j=j>>4;
+	if (j<10) {
+		if (ptr[i+1] != '0'+j) return 0;
+	} else {
+		if (ptr[i+1] != 'A'+j-10) return 0;
+	}
+	return i;
+}
+
+void checksum (char *ptr) {
+	int i,j,k;
+
+	for (i=1,j=0,k=strlen(ptr);i<k;++i) j^=ptr[i];
+	sprintf(ptr+k,"*%02X",j);
+}
+
```

Comment:
2 ways to calculate the checksum.

Reply:
checksum() generates a checksum and adds it to the end of the sentence.
verify_checksum() validates that the checksum at the end of the sentence is correct.  They serve two different purposes and have only the for loop mostly in common.